### PR TITLE
Generate index for objective-c code. Remap and move index to deriveddata.

### PR DESCRIPTION
**Pull request by amberdixon**
This PR adds the `index-store-path` option when invoking clang to compile objective-c to build libraries. It also adds a step to the post-build installer which uses index-import to update the paths referenced by the index generated by bazel and place the index in the derived data directory.

I set the `CLANG_ENABLE_MODULES` setting to true, because in local development, I noticed that indexing was failing on some of my projects, due to this build setting not being set.

I also made the operations performed in the installer script write to log files in a new bazel-diagnostics directory that lives in derived data. This should make it easier to diagnose issues.

Other notes:
I called `defaults write com.apple.dt.Xcode IDEIndexShowLog  YES` from Terminal and then was able to view what xcode's indexer was doing. Note that if Xcode's indexer isn't enabled or if the indexer fails when it runs clang on a source file, Xcode auto-complete functionality won't work! We still need to figure out a way to convince Xcode that it's indexer is working, so that autocomplete operations will work on the bazel generated index.

When the sample code that is used for testing in this repo is compiled, no index records or units are generated. Therefore, I did not test my implementation against this sample code. (In the future, we should tweak the sample code so that it will generate index files. This may involve adding some methods/classes/a public API to it.) 

When I tested locally, I pointed my own iOS repo to my local rules_ios checkout.
1. Generate an xcodeproject using `bazel run :ProjectName`.
2. Clear the derived data Index/DataStore directory for this xcodeproject if one already exists.
3. Disable Xcode's indexer by running `defaults write com.apple.dt.XCode IDEIndexDisable 1` from terminal. This indexer will generate record and unit files (if it works) and then you won't necessarily be using bazel-generated record and unit files.
4. Open the project in Xcode and build locally.
5. Go into your derived data directory. Optionally:
 a. Check out the index-import library and build it. The latest version on master includes a binary called validate-index that ensures your unit paths are valid. Run validate-index on DerivedDataForYourProject/Index/DataStore. The command should return an exit status of 0.
 b. Run absolute-unit on one of the unit files in DerivedDataForYourProject/Index/DataStore/v5/units to sanity check that units are being generated and mapped correctly.

### CODE REVIEW

File: rules/library.bzl [hunk 1]
```diff
@@ -568,6 +568,7 @@ def apple_library(name, library_tools = {}, export_private_headers = True, names
 
     objc_library_data = library_tools["wrap_resources_in_filegroup"](name = objc_libname + "_data", srcs = data)
     objc_copts.append("-I.")
+    objc_copts.extend(("-index-store-path", "$BAZEL_WORKSPACE_DIR/bazel-rules-ios-apple-library-objc-indexstores/apple-library-objc.indexstore"))
```
[the remark this answers is not in the document]
Comment (amberdixon) [reply rules/library.bzl:571]: I'll switch it to use GENDIR.

File: tools/xcodeproj-shims/installers/_indexstore.sh [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+# Example: /private/var/tmp/_bazel_<username>/<hash>/execroot/<workspacename>
+readonly bazel_root="^/private/var/tmp/_bazel_.+?/.+?/execroot/[^/]+"
+readonly bazel_bin="^(?:$bazel_root/)?bazel-out/.+?/bin"
+
+# Object file paths are fundamental to how Xcode loads from the indexstore. If
+# the `OutputFile` does not match what Xcode looks for, then those parts of the
+# indexstore will not be found and used.
+readonly bazel_swift_object="$bazel_bin/.*/(.+?)(?:_swift)?(?:[.]library)?_objs/.*/(.+?)[.]swift[.]o$"
+readonly bazel_objc_object="$bazel_bin/.*/(?:[.]library)?_objs/(?:arc/|non_arc/)?(.+?)-(?:objc|cpp)/(.+?)[.]o$"
```
Comment (kastiglione) [inline tools/xcodeproj-shims/installers/_indexstore.sh:20]: The `.library` suffix is a convention we use in some places at Lyft. If you don't use a `.library` suffix in any of your macros/rules, then you can remove the `(?:[.]library)?` portions of these regex.

File: tools/xcodeproj-shims/installers/_indexstore.sh [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+readonly bazel_module="$bazel_bin/.*/(.+?)[.]swiftmodule$"
+readonly xcode_module="$BUILT_PRODUCTS_DIR/\$1.swiftmodule/$ARCHS.swiftmodule"
+
+# External dependencies are available via the `bazel-<workspacename>` symlink.
+# This remapping, in combination with `xcode-index-preferences.json`, enables
+# index features for external dependencies, such as jump-to-definition.
+readonly bazel_external="$bazel_root/external"
+readonly xcode_external="$BAZEL_WORKSPACE_ROOT/bazel-$(basename "$SRCROOT")/external"
+
+$BAZEL_PATH run @com_github_lyft_index_import//:index_import \
```
[the remark this answers is not in the document]
Comment (amberdixon) [reply tools/xcodeproj-shims/installers/_indexstore.sh:34]: It's 3.6MB. Would that make the time it takes to git pull for this repo annoying long?

File: tools/xcodeproj-shims/installers/indexstores.sh [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+echo "Start remapping index files"
+
+# Make sure add these to bazel build target copts for objective-c
+find "$BAZEL_WORKSPACE_ROOT/bazel-out/*/bin/" \
+     -type d \
+     -name "*.indexstore" \
+     -print0 \
+    | xargs -0 "$BAZEL_INSTALLERS_DIR/_indexstore.sh"
+
+"$BAZEL_INSTALLERS_DIR/_indexstore.sh" $BAZEL_WORKSPACE_DIR/bazel-rules-ios-apple-library-objc-indexstores/apple-library-objc.indexstore
```
[the remark this answers is not in the document]
Comment (amberdixon) [reply tools/xcodeproj-shims/installers/indexstores.sh:15]: When compilation doesn't happen locally, the index store path doesn't get regenerated :( :( Not sure if there's a simple solution here.

File: tools/xcodeproj-shims/installers/_indexstore.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (segiddins) [reply tools/xcodeproj-shims/installers/_indexstore.sh:34]: We can copy it using the installer, it wouldn’t require pulling the binary into this repo

File: tools/xcodeproj-shims/installers/indexstores.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (amberdixon) [reply tools/xcodeproj-shims/installers/indexstores.sh:15]: I'll file an issue to figure this out. It probably should not prevent landing this PR.

File: rules/library.bzl (hunk 1 above)
[the remark this answers is not in the document]
Comment (amberdixon) [reply rules/library.bzl]: FYI to reviewers, the usage of GENDIR is explained here. Search the page for "GENDIR" https://github.com/bazelbuild/bazel/blob/master/site/docs/output_directories.md

File: tools/xcodeproj-shims/installers/_indexstore.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (amberdixon) [reply tools/xcodeproj-shims/installers/_indexstore.sh]: Updated!

File: tools/xcodeproj-shims/installers/_indexstore.sh (hunk 2 above)
Comment (amberdixon) [reply tools/xcodeproj-shims/installers/_indexstore.sh]: Thanks! I removed that part of the regex.

File: tests/macos/xcodeproj/Single-Application-Project-AllTargets.xcodeproj/bazelinstallers/_swiftmodule.sh [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+
+    cp "$module" "$module_bundle/$NATIVE_ARCH_ACTUAL.swiftmodule"
+    cp "$doc" "$module_bundle/$NATIVE_ARCH_ACTUAL.swiftdoc"
+
+    ios_module_name="$NATIVE_ARCH_ACTUAL-$LLVM_TARGET_TRIPLE_VENDOR-$SWIFT_PLATFORM_TARGET_PREFIX$LLVM_TARGET_TRIPLE_SUFFIX"
+    cp "$module" "$module_bundle/$ios_module_name.swiftmodule"
+    cp "$doc" "$module_bundle/$ios_module_name.swiftdoc"
+
+    chmod -R +w "$module_bundle"
+done
```
Comment (segiddins) [inline tests/macos/xcodeproj/Single-Application-Project-AllTargets.xcodeproj/bazelinstallers/_swiftmodule.sh:19]: nit: missing newline

File: tests/macos/xcodeproj/Single-Application-Project-DirectTargetsOnly.xcodeproj/bazelinstallers/_indexstore.sh [hunk 6]
```diff
@@ -0,0 +1,43 @@
+#!/bin/bash
+
+set -euo pipefail
+echo "running index import"
```
Comment (segiddins) [inline tests/macos/xcodeproj/Single-Application-Project-DirectTargetsOnly.xcodeproj/bazelinstallers/_indexstore.sh:4]: debugging echo?

File: tests/macos/xcodeproj/Single-Application-Project-DirectTargetsOnly.xcodeproj/bazelinstallers/_swiftmodule.sh [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+
+    cp "$module" "$module_bundle/$NATIVE_ARCH_ACTUAL.swiftmodule"
+    cp "$doc" "$module_bundle/$NATIVE_ARCH_ACTUAL.swiftdoc"
+
+    ios_module_name="$NATIVE_ARCH_ACTUAL-$LLVM_TARGET_TRIPLE_VENDOR-$SWIFT_PLATFORM_TARGET_PREFIX$LLVM_TARGET_TRIPLE_SUFFIX"
+    cp "$module" "$module_bundle/$ios_module_name.swiftmodule"
+    cp "$doc" "$module_bundle/$ios_module_name.swiftdoc"
+
+    chmod -R +w "$module_bundle"
+done
```
Comment (segiddins) [inline tests/macos/xcodeproj/Single-Application-Project-DirectTargetsOnly.xcodeproj/bazelinstallers/_swiftmodule.sh:19]: nit: trailing newline

File: tools/xcodeproj_shims/install.sh [hunk 8]
```diff
@@ -10,6 +10,13 @@
 set -eux
 
 
+diagnostics_dir="$TARGET_BUILD_DIR/../../../bazel-xcode-diagnostics/"
```
Comment (segiddins) [inline tools/xcodeproj_shims/install.sh:13]: will the files in here ever get cleaned up?

File: tools/xcodeproj_shims/installers/swiftmodules.sh [hunk 9]
```diff
@@ -4,13 +4,8 @@ set -euo pipefail
...
 # Copy Bazel build `.swiftmodule` files to `DerivedData`. This is used by Xcode
 # and its indexing.
-
-# TODO: what should BAZEL_BIN be?
-# something like this:
-# readonly config=macos-x86_64-min10.13-applebin_macos-darwin_x86_64-dbg
-# export BAZEL_BIN=bazel-out/$config/bin
-find "$BAZEL_BIN" \
-     -name "*.swiftmodule" \
+find bazel-out/*/bin/ \
```
Comment (segiddins) [inline tools/xcodeproj_shims/installers/swiftmodules.sh:7]: should this be `$BAZEL_WORKSPACE_ROOT/bazel-out` ?

File: tools/xcodeproj_shims/installers/indexstores.sh [hunk 10]
```diff
@@ -0,0 +1,14 @@
+#!/bin/bash
+
+set -euo pipefail
+
+# See `_indexstore.sh` for full details.
+
+# Make sure add these to bazel build target copts for objective-c
+find bazel-out/*/bin/ \
```
Comment (segiddins) [inline tools/xcodeproj_shims/installers/indexstores.sh:8]:
```suggestion
find "$BAZEL_WORKSPACE_ROOT/bazel-out/"*/bin/ \
```

File: tools/xcodeproj_shims/install.sh (hunk 8 above)
Comment (amberdixon) [reply tools/xcodeproj_shims/install.sh:13]: I'll add code to delete older files and generate a new file for each run that includes the current timestamp in the filename.

**Pull request merged**